### ISSUE TITLE
stopforumspam doesn't support ipv6 addresses

### ISSUE THREAD

**Issue by psychobunny** [NodeBB/NodeBB#3465]
Which blows up your forum if they entered the registration queue

**Comment by psychobunny** [comment]
reopening, need to backport to 0.7.3

**Comment by damianb** [comment]
of note, this should be ready soon: http://www.stopforumspam.com/forum/viewtopic.php?id=6392

as an aside, you should try using the confidence scoring data (or passing it along to the admin) - pass &confidence param as a trigger.  http://www.stopforumspam.com/forum/viewtopic.php?id=3941

algo was passed to pedigree for implementation by yours truly. :)

**Comment by barisusakli** [comment]
I actually added the frequency, appears and confidence values into the tooltips here https://github.com/NodeBB/NodeBB/blob/v0.7.x/src/views/admin/manage/registration.tpl#L18

**Comment by damianb** [comment]
awesome to see, thanks @barisusakli. not sure if it'll matter to you, but confidence is a percentage, whereas frequency is (I believe) appearances within the last year.

**Comment by julianlam** [comment]
That's good to know, since I didn't actually have any idea what those numbers meant...

**Comment by damianb** [comment]
@julianlam confidence is based on that algo stated in the post - over time, confidence decays unless the data point (ip/username/email) is seen again, and the more often the data point has been seen in the last year (the frequency), the slower and less it decays.  I don't believe I have the algo lying around anywhere easily findable atm, but going by memory - say you had an email appear 10 times in a year.  It'd be (ballparking this) like sixty percent near the beginning, decaying rapidly until sitting around twenty, ten percent or less until all results drop off the API.  Now, if that same email appeared 5000 times in a year, it'd be scoring something like ninty percent near the beginning, decaying to about...let's say fifty percent near the end.  

tl;dr: The algo itself tries to approximate how likely the data point is to be a spammer based on reporting frequency and the last report date.  

 * If it's barely reported or it's been a long time since it was reported, confidence is low.  
 * If it's been oft reported or it's been reported recently confidence is moderate.
 * If it's been reported a ton **and** reported recently, confidence is high. 

---

Sidebar: Since username is unreliable and has significant conflicts with reality, usually I found that to be worth ignoring (only really relevant when username is reported in combination with the same email or IP).  Email and IP are much more reliable on their own though, since that is fairly consistent with most spambots.

Might also be worth hardcoding gmail dot expansion when checking to see if an email address is already in use - gmail's policy of ignoring dots is oft-abused by spambots that register and reregister over and over, often using the same email to build a horde of accounts.

**Comment by psychobunny** [comment]
@barisusakli @julianlam, we should add a new setting to put users into the registration queue only if the confidence score is beyond a certain threshold (settable in acp)

Instead of putting everybody in the queue. @damianb what would your recommended value look like?

**Comment by julianlam** [comment]
yeah perhaps a slider with a recommended confidence value. 

Thing is we've seen lots of spammers from Korea (lately) that pass all of the tests. 

Any effort to cull the registration queue automatically while minimizing false positives is a good idea in my books.

**Comment by damianb** [comment]
@psychobunny 

* Username >= 85-90% (again, because unreliable; high confidence should limit false positive risk. may be best left ignored)
* Email >= 50% (emails rarely go from a spammer *back* to a legit user)
* IP >= 60-70% (will likely require tuning, possibly backing it up slightly higher - IPs can frequently change hands from an illegitimate user to a legit one)

@julianlam Any patterns in that traffic? Any way to correlate patterns - common useragents from logs, form submission times, etc.?  I have a few tricks up my sleeve that may still help.

**Comment by psychobunny** [comment]
Awesome. Thanks!

**Comment by damianb** [comment]
@psychobunny may also be worth tracking registration "duration" - having that data point can be a good determinator of human/not human (assuming standard email registration, not third-party signin).

Recording the registration UA is also sometimes useful; some bots like to use stupid ones such as IE6, IE7, obviously incompatible stuff. Years ago, they'd also use "Windows XP", which was obviously fake and a big red flag (WinXP identified as Win NT 5.1 in the UA).

Really, I would suggest adding in some points to datamine the registration process and see what the real problem users are doing.  

Good questions to ask yourselves: 

* Are their browsers requesting asset files?  
* Are they hitting the registration page straight, or following links themselves?
* How long does it take for them to register after their first visit to the site?
* If a CAPTCHA is in use, how many times is it being attempted before succeeding?

Also a good idea is to provide a way to contact the site admins directly, in case of issues during registration - some disabled users *cannot* get through normal registration processes, and if you allow yourself the capability to assist them with that, you have less registration turnover.

**Comment by damianb** [comment]
Also, sorry for the massive braindump. Been a while since I've been involved in the antispam stuff - things are floating to the surface, and I figure it'd be useful to get it down. :)

**Comment by psychobunny** [comment]
I was going to ask, do you do this for a living? Thanks for all the info, we need to dissect this into several sub issues

**Comment by damianb** [comment]
Used to be heavily involved with the antispam efforts with stopforumspam a few years ago, both as a contributing user and then as a moderator there, before I hit a time crunch and had to narrow my focus (since then, I've moved on to other things & places, such as my current mod role on /r/netsec).  Just noticed this commit as I was browsing, though, and thought I'd chime in; been watching nodebb off and on ever since you guys started getting things together, and have been pleased to see it not get stuck in the same rut that phpBB had/has.

**Comment by psychobunny** [comment]
Cool. Yeah I remember you posted a bunch of issues when we were starting up pointing out things in the code base that needed refactoring. Helped a lot. Thanks :)
